addons/sale_coupon/models/sale_order.py

```diff
@@ -375,13 +414,17 @@ def update_line(order, lines, values):
                             value_found = True
                             # Working on Case 3.
                             lines_to_remove -= line
+                            lines_to_keep += [(0, False, value)]
                             lines_to_remove += update_line(order, line, value)
                             continue
                     # Case 2.
                     if not value_found:
-                        order.write({'order_line': [(0, False, value)]})
+                        lines_to_add += [(0, False, value)]
                 # Case 3.
-                lines_to_remove.unlink()
+                if lines_to_remove:
+                    lines_to_remove.unlink()
+                    order.write({'order_line': lines_to_keep})
+                order.write({'order_line': lines_to_add})
```

Comment:
Out of scope for this PR but couldn't these be done in a single write? Using `Command.UNLINK`

Reply:
I didn't think about it before but I'll test doing it :D

Reply:
Nice catch, could have been part of a separate PR tho, to not noise the commits as those are already quite touchy :)